sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/FluxInputStream.java

```diff
@@ -0,0 +1,242 @@
+        if (this.buffer.available() > 0) {
+            return this.buffer.read(b, off, len);
+        }
+
+        /* If the flux completed, there is no more data available to be read from the stream. Return -1. */
+        if (this.fluxComplete) {
+            return -1;
+        } else {
+            throw logger.logExceptionAsError(new IllegalStateException("An unexpected error occurred. No data was "
+                + "read from the stream but the stream did not indicate completion."));
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        subscription.cancel();
+        if (this.buffer != null) {
+            this.buffer.close();
+        }
+        super.close();
+        if (this.lastError != null) {
+            throw logger.logThrowableAsError(this.lastError);
+        }
+    }
+
+    /**
+     * Request more data and wait on data to become available.
+     */
+    private void blockForData() {
+        lock.lock();
+        try {
+            waitingForData = true;
+            if (!subscribed) {
+                subscribeToData();
+            } else {
+                subscription.request(1);
+            }
+            // Block current thread until data is available.
+            while (waitingForData) {
+                if (fluxComplete) {
+                    break;
+                } else {
+                    try {
+                        dataAvailable.await();
+                    } catch (InterruptedException e) {
+                        throw logger.logExceptionAsError(new RuntimeException(e));
+                    }
+                }
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Subscribes to the data with a special subscriber.
+     */
+    private void subscribeToData() {
+        this.data
+            .onBackpressureBuffer(Constants.GB)
+            .subscribe(
+                // ByteBuffer consumer
+                byteBuffer -> {
+                    lock.lock();
+                    try {
+                        this.buffer = new ByteArrayInputStream(FluxUtil.byteBufferToArray(byteBuffer));
+                        this.waitingForData = false;
+                        // Signal the consumer when data is available.
+                        dataAvailable.signal();
+                    } finally {
+                        lock.unlock();
+                    }
+                },
+                // Error consumer
+                throwable -> {
+                    // Signal the consumer in case an error occurs (indicates we completed without data).
+                    signalOnCompleteOrError();
+                    if (throwable instanceof HttpResponseException) {
+                        this.lastError = new IOException(throwable);
+                    } else if (throwable instanceof IllegalArgumentException) {
+                        this.lastError = new IOException(throwable);
```

Comment:
I don't think we need to wrap IllegalArgumentException in an IOException. And I don't think we need to wrap ioExceptions in another IOException.

Reply:
at least for IllegalArgument - is the read API allowed to throw anything other than IOexception? I'm not certain

Reply:
IllegalArgumentException sounds like validation failed somewhere. That usually should bubble up. Isn't it runtimeexception?

Reply:
Here actually should I just wrap any type of exception except IOException in an IOEXception? I'm forgetting why I even checked for specific types of throwables.

Reply:
Like this 
                    if (throwable instanceof IOException) {
                        this.lastError = (IOException) throwable;
                    } else {
                        this.lastError = new IOException(throwable);
                    }

Reply:
I don't think we want to wrap IllegalArgumentException, etc. Probably shouldn't wrap any RuntimeExceptions, so rethrow IOExceptions and RuntimeExceptions, wrap HttpResponseExceptions. How does that sound?

Reply:
If we do that exactly though we would be silently ignoring any other Exception which would be really bad too

Reply:
https://stackoverflow.com/questions/52195112/what-is-a-ioexception-and-how-do-i-fix-it/52195182
https://docs.oracle.com/javase/7/docs/api/java/io/IOException.html
https://stackoverflow.com/questions/13216148/java-what-throws-an-ioexception

The IOException is quite ambiguous. Do we know what are scenarios when RuntimeException is thrown here?

Reply:
I think it's dependent on the behavior of the Flux. In our primary use case, the Flux we care about is the one we return in blob.query - the code that handles that would take in the inputs, and it could return some type of error if the query failed (that would be IOException), or the input validation failed (IllegalArgument), etc

Reply:
Summarizing our offline discussion. Anything thrown during pulling data is legit IOException. For IllegalArgument and it's family we should find a way to invoke validation eagerly (i.e. these exceptions should come from `query`/`queryWithResponse` and shouldn't be deferred until someone tries to read from stream).

Reply:
Resolveed by moving input validation further up by blocking on the network request